core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -423,6 +462,20 @@ default void onSuccess() {
          */
         default void onFailure(final Throwable t) {
         }
+
+    }
+
+    public static class IllegalShardFailureException extends ElasticsearchException {
```

Comment:
It's a shame we add a new exception here. Is there nothing we can reuse? if we end up adding something, can we name it explicitly to what we check -  "NoLongerPrimaryException"?